src/wp-includes/class-wp-theme-json.php

```diff
@@ -0,0 +1,418 @@
+		$schema['settings']['blocks'] = $schema_settings_blocks;
+
+		// Remove anything that's not present in the schema.
+		foreach ( array( 'settings' ) as $subtree ) {
+			if ( ! isset( $input[ $subtree ] ) ) {
+				continue;
+			}
+
+			if ( ! is_array( $input[ $subtree ] ) ) {
+				unset( $output[ $subtree ] );
+				continue;
+			}
+
+			$result = self::remove_keys_not_in_schema( $input[ $subtree ], $schema[ $subtree ] );
+
+			if ( empty( $result ) ) {
+				unset( $output[ $subtree ] );
+			} else {
+				$output[ $subtree ] = $result;
+			}
+		}
+
+		return $output;
+	}
+
+	/**
+	 * Given a tree, removes the keys that are not present in the schema.
+	 *
+	 * It is recursive and modifies the input in-place.
+	 *
+	 * @param array $tree Input to process.
+	 * @param array $schema Schema to adhere to.
+	 *
+	 * @return array Returns the modified $tree.
+	 */
+	private static function remove_keys_not_in_schema( $tree, $schema ) {
+		$tree = array_intersect_key( $tree, $schema );
+
+		foreach ( $schema as $key => $data ) {
+			if ( ! isset( $tree[ $key ] ) ) {
+				continue;
+			}
+
+			if ( is_array( $schema[ $key ] ) && is_array( $tree[ $key ] ) ) {
+				$tree[ $key ] = self::remove_keys_not_in_schema( $tree[ $key ], $schema[ $key ] );
+
+				if ( empty( $tree[ $key ] ) ) {
+					unset( $tree[ $key ] );
+				}
+			} elseif ( is_array( $schema[ $key ] ) && ! is_array( $tree[ $key ] ) ) {
+				unset( $tree[ $key ] );
+			}
+		}
+
+		return $tree;
+	}
+
+	/**
+	 * Returns the existing settings for each block.
+	 *
+	 * Example:
+	 *
+	 * {
+	 *   'root': {
+	 *     'color': {
+	 *       'custom': true
+	 *     }
+	 *   },
+	 *   'core/paragraph': {
+	 *     'spacing': {
+	 *       'customPadding': true
+	 *     }
+	 *   }
+	 * }
+	 *
+	 * @return array Settings per block.
+	 */
+	public function get_settings() {
+		if ( ! isset( $this->theme_json['settings'] ) ) {
+			return array();
+		} else {
+			return $this->theme_json['settings'];
+		}
+	}
+
+	/**
+	 * Builds metadata for the setting nodes, which returns in the form of:
+	 *
+	 * [
+	 *   [
+	 *     'path' => ['path', 'to', 'some', 'node' ]
+	 *   ],
+	 *   [
+	 *     'path' => [ 'path', 'to', 'other', 'node' ]
+	 *   ],
+	 * ]
+	 *
+	 * @param array $theme_json The tree to extract setting nodes from.
+	 *
+	 * @return array
+	 */
+	private static function get_setting_nodes( $theme_json ) {
+		$nodes = array();
+		if ( ! isset( $theme_json['settings'] ) ) {
+			return $nodes;
+		}
+
+		// Top-level.
+		$nodes[] = array(
+			'path' => array( 'settings' ),
+		);
+
+		// Calculate paths for blocks.
+		if ( ! isset( $theme_json['settings']['blocks'] ) ) {
+			return $nodes;
+		}
+
+		foreach ( $theme_json['settings']['blocks'] as $name => $node ) {
+			$nodes[] = array(
+				'path' => array( 'settings', 'blocks', $name ),
+			);
+		}
+
+		return $nodes;
+	}
+
+	/**
+	 * Merge new incoming data.
+	 *
+	 * @param WP_Theme_JSON $incoming Data to merge.
+	 */
+	public function merge( $incoming ) {
```

Comment:
Note that this function hasn't been updated to include the changes to the merge algorithm introduced at https://github.com/WordPress/gutenberg/pull/31669 It needs the 10.7 Gutenberg packages to work as expected. It should be updated later.